megamek/src/megamek/common/weapons/MissileWeaponHandler.java

```diff
@@ -485,55 +486,121 @@ protected int getAMSHitsMod(Vector<Report> vPhaseReport) {
-                                apdsEnt.getEquipmentNum(counter),
-                                entityTarget);
-                    }
-                    if (!(counter.getType() instanceof WeaponType)
-                            || !counter.isReady() || counter.isMissing()
-                            // no AMS when a shield in the AMS location
-                            || (apdsEnt.hasShield() && apdsEnt.hasActiveShield(
-                                    counter.getLocation(), false))
-                            // shutdown means no AMS
-                            || apdsEnt.isShutDown()
-                            // AMS only fires vs attacks in arc covered by ams
-                            || !isInArc) {
-                        continue;
-                    }
+                boolean isAMSBay = counter.getType().hasFlag(WeaponType.F_AMSBAY);
+                boolean isAPDS = counter.isAPDS();
+                
+                //Only one AMS and one APDS can engage each missile attack
+                if (isAMS && amsEngaged) {
+                    continue;
+                }
+                if (isAPDS && apdsEngaged) {
+                    continue;
+                }
+                
+                //Check the firing arc, even though this was done when the AMS was assigned
+                Entity pdEnt = counter.getEntity();
+                boolean isInArc;
+                // If the defending unit is the target, use attacker for arc
+                if (entityTarget.equals(pdEnt)) {
+                    isInArc = Compute.isInArc(game, entityTarget.getId(),
+                            entityTarget.getEquipmentNum(counter),
+                            ae);
+                } else { // Otherwise, the attack target must be in arc
+                    isInArc = Compute.isInArc(game, pdEnt.getId(),
+                            pdEnt.getEquipmentNum(counter),
+                            entityTarget);
+                }
+                
+                if (!isInArc) {
+                    continue;
+                }
+                
+                // Point defenses can't fire if they're not ready for any other reason
+                if (!(counter.getType() instanceof WeaponType)
+                        || !counter.isReady() || counter.isMissing()
+                        // no AMS when a shield in the AMS location
+                        || (pdEnt.hasShield() && pdEnt.hasActiveShield(
+                                counter.getLocation(), false))
+                        // shutdown means no AMS
+                        || pdEnt.isShutDown()) {
+                    continue;
+                }
+                
+                //If we're an AMSBay, heat and ammo must be calculated differently
+                if (isAMSBay) {
+                    for (int wId : counter.getBayWeapons()) {
+                        Mounted bayW = entityTarget.getEquipment(wId);
```

Comment:
I seem to recall some kind of function you wrote for a previous PR that handles AMS activity. Is there a way to reuse any of that code?

Reply:
No, though I did use it as a reference for streamlining this method.  I had to add AMSBays as a possible counter for standard cluster table missile attacks, and the code here breaks the bay down and assigns 1 AMS in it to each attack rather than combining the entire bay - which is incompatible with the damage and heat code for bays in calcCounterAV().